lib/chef/knife/cookbook_upload.rb

```diff
@@ -202,15 +205,46 @@ def assert_environment_valid!
       end
 
       def upload(cookbooks, justify_width)
-        cookbooks.each do |cb|
-          ui.info("Uploading #{cb.name.to_s.ljust(justify_width + 10)} [#{cb.version}]")
-          check_for_broken_links!(cb)
-          check_for_dependencies!(cb)
+        require "tmpdir"
+
+        Dir.mktmpdir do |tmp_dir_path|
+          begin
+            cookbooks_to_upload = []
+            compiled_metadata = nil
+
+            cookbooks.each do |cb|
+              compiled_metadata = cb.compile_metadata
+              ui.info("Uploading #{cb.name.to_s.ljust(justify_width + 10)} [#{cb.version}]")
+              check_for_broken_links!(cb)
+              check_for_dependencies!(cb)
+
+              if compiled_metadata
+                proxy_cookbook_path = "#{tmp_dir_path}/#{cb.name}"
+                # Make a symlink
+                file_class.symlink cb.root_dir, proxy_cookbook_path
+                proxy_loader = Chef::Cookbook::CookbookVersionLoader.new(proxy_cookbook_path)
+
+                proxy_loader.load_cookbooks
+                cookbook_to_upload = proxy_loader.cookbook_version
+                cookbook_to_upload.freeze_version if options[:freeze]
+                cookbooks_to_upload << cookbook_to_upload
```

Comment:
If you're trying to implement my suggestion to do a tmpdir what that should probably look more like is:

```
  create the temp dir
  loop over all the cookbooks
    copy whole cookbook to "#{tmp_dir_path}/{cb.name}"  # do for every one
    use CookbookVersionLoader to build the CookbookVersion object
    call compile_metadata on the CookbookVersion object
    add to the cookbooks_to_upload list
  end
  upload all the cookbooks in the cookbooks_to_upload list
  <tempdir is deleted when it goes out of scope along with all the metadata.json>
```

if you do that, then you don't need the ensure/GC.start/File.unlink magic.

there's some excessive copying that isn't strictly necessary here, but i wouldn't worry about trying to speed it up, this isn't very performance critical code, and the upload across the network will take a lot longer than pushing files around on disk.

Reply:
I suspect that a much cleaner way to write this would be:

```
# Chef::CookbookLoader.working_directory should be a factory method that produces a 
# Chef::CookbookLoader object after creating a tmpdir and copying all the cookbooks in the
# Hash passed to it here.  The tmpdir object needs to be saved on the CookbookLoader to
# prevent it from being deleted when the method exits (`saved_tmpdir` accessor)
cl = Chef::CookbookLoader.working_directory_from_hash(cookbooks)
# build the internal CookbookVersions
cl.load_cookbooks
# call compile_metadata on all CookbookVersions
cl.compile_metadata
# freeze all the versions in the collection
cl.freeze_versions if options[:freeze]
# upload all the versions
cl.upload(force: config[:force], concurrency: config[:concurrency])
# FileUtils.rm_rf the saved_tmpdir.path or throw and internal error if that is nil (doesn't have to be a pretty exception since it would mean a codebug
cl.unlink!
```

Reply:
Okay @lamont-granquist will have a look at this and update you accordingly. Thank you :)

Reply:
@lamont-granquist Are you proposing here to refactor the current code at line https://github.com/chef/chef/blob/master/lib/chef/knife/cookbook_upload.rb#L96-L142 ?

As per the current code if user select to upload all cookbooks, there is a single call to upload method. https://github.com/chef/chef/blob/master/lib/chef/knife/cookbook_upload.rb#L106

But in case of user selects specific cookbooks to upload then each cookbook gets uploaded separately https://github.com/chef/chef/blob/master/lib/chef/knife/cookbook_upload.rb#L116 and the progress is shown using upload_ok and upload_failure flags.

Reply:
For the upload-everything use case you can probably do this instead of Chef::CookbookLoader.working_directory_from_hash:

```
temp_cl = Chef::CookbookLoader.working_directory_from_cl(cl)
```

So that routine should create a tmpdir and then copy all of the cookbooks in the specified cookbook_loader object to a new cookbook_loader object that references the tmpdir.

Reply:
For the uploads I see where we have to upload them one-by-one and knife has to do its own error handling per-cookbook so the best that could be done would be this code:

```
          cookbook_repo.each do |cookbook_name, cookbook|
            cookbooks_for_upload << cookbook
            cookbook.freeze_version if config[:freeze]
            version_constraints_to_update[cookbook_name] = cookbook.version
          end
```

could look like:

```
          temp_cl = Chef::CookbookLoader.working_directory_from_cl(cookbook_repo)
          temp_cl.freeze_versions if config[:freeze]
          temp_cl.each do |cookbook_name, cookbook|
            cookbooks_for_upload << cookbook
            version_constraints_to_update[cookbook_name] = cookbook.version
          end
          [ .. and then call upload on every cookbook ... ]
```